### PR TITLE
fix the bug that connection cannot be closed when the client close request to /containers/{stopped container id}/stats

### PR DESCRIPTION
**Pull request by chengjingtao**
`Bug description`:
when client connect to swarm ,and request to `/containers/{ id of stopped container }/stats` api .the docker server didn't write any data in body to swarm. But when we close the request of client, the connection of swarm and docker server was not be closed.

I have try to fix the bug .

`Reproduce`:
As I say ,you can reproduce the bug : 
1. Join your docker host into swarm
2. Run a container and stop it  on your host (Don't forget to stop it !)
3. Using your browser, and request to  http://swarm-address/containers/{container you just stopped}/stats . and wait for a seconds and stop it .  
4. Do `step 3` for many times and exec `sudo lsof -i:2375` ,  you will see the connection created again and again.  

Signed-off-by:  chengjt chengjt@csdn.net

### CODE REVIEW

**Comment by MHBauer** [comment]
quick notes.
1. gofmt 
2. change all instances of `dected` to `detected`

**Comment by linecheng** [comment]
Ok

File: api/utils.go [hunk 1]
```diff
@@ -103,6 +103,48 @@ func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request,
...
+			closeNotifier http.CloseNotifier
+			ok            bool
+		)
+
+		type canceler interface {
+			CancelRequest(*http.Request)
+		}
+
+		if closeNotifier, ok = w.(http.CloseNotifier); ok == false {
+			log.Error("ResponseWriter was not support CloseNotify")
```
Comment (MHBauer) [inline api/utils.go]: was -> does

File: api/utils.go [hunk 2]
```diff
@@ -103,6 +103,48 @@ func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request,
...
+		var (
+			closeNotifier http.CloseNotifier
+			ok            bool
+		)
+
+		type canceler interface {
+			CancelRequest(*http.Request)
+		}
+
+		if closeNotifier, ok = w.(http.CloseNotifier); ok == false {
```
Comment (MHBauer) [inline api/utils.go]: usually I see `!ok` instead of `ok == false`. Style thing, but I think `!ok` is more idiomatic.

File: api/utils.go [hunk 3]
```diff
@@ -103,6 +103,48 @@ func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request,
...
+			log.Error("ResponseWriter was not support CloseNotify")
+			return
+		}
+
+		select {
+		case <-closeNotifier.CloseNotify():
+			{
+				var cancel canceler
+				if cancel, ok = client.Transport.(canceler); ok == false {
```
Comment (MHBauer) [inline api/utils.go]: same as above. `!ok`.

File: api/utils.go [hunk 4]
```diff
@@ -103,6 +103,48 @@ func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request,
...
+					log.Error("client.Transport was not support CancelRequest")
```
Comment (MHBauer) [inline api/utils.go]: was -> does

File: api/utils.go [hunk 5]
```diff
@@ -103,6 +103,48 @@ func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request,
...
+					return
+				}
+
+				cancel.CancelRequest(r)
```
Comment (MHBauer) [inline api/utils.go]: Looking at golang docs I see: `Deprecated: Use Request.Cancel instead.`
Not sure how strict we need to be on enforcing this.
I think the replacement is `close(r.Cancel)`.

**Comment by MHBauer** [comment]
Please squash the different commits after you make changes.

**Comment by linecheng** [comment]
@MHBauer  thanks for your review.
I fixed some grammatical errors .
you are right , `Transport.CancelRequest(r)` is deprecated.
I have try to use "close(r.Cancel)"  //r.Cancel is receive only.
```

ch := make(chan struct{})
r.Cancel=ch
go func(){

     ...
     ...
     close(ch)   // cancel.CancelRequest(r)
     ...
     ...

}()

```

But  I find that my code dose not work  well suddenly
Neither using the old `cancel.CancelRequest(r)` nor `close(r.Cancel)` can cancel the request really, I'm confused about my previous tests.

I am not sure about the reason of that. It seams to be related to `client, scheme, err := engine.HTTPClientAndScheme()`. 
I am checking it and  will squash the different commits  after checking  .

Thanks again.

**Pull request closed without merging**